### PR TITLE
Updating distributed training guide

### PR DESCRIPTION
**Pull request by mx-iao**
# PR into Azure/azureml-web

## Checklist

I have:

- [x] read and followed the contributing guidelines

## Changes

- update distributed training guide
- remove mpi workaround and add per-process and per-node PyTorch distributed training docs

addresses #31

### CODE REVIEW

File: website/docs/cheatsheet/distributed-training.md [hunk 1]
```diff
@@ -6,524 +6,335 @@ keywords:
...
-trainer.fit(model, train_loader, val_loader)
+Hugging Face provides many [examples](https://github.com/huggingface/transformers/tree/master/examples) for using its Transformers library with `torch.distributed.launch` to run distributed training. To run these examples and your own custom training scripts using the Transformers Trainer API, follow the [Using `torch.distributed.launch`](#using-torchdistributedlaunch-per-node-launch) section.
+
+Sample job configuration code to fine-tune the BERT large model on the text classification MNLI task using the `run_glue.py` script on one node with 8 GPUs:
+```python
+from azureml.core import ScriptRunConfig
+from azureml.core.runconfig import PyTorchConfiguration
+
+distr_config = PyTorchConfiguration() # node_count defaults to 1
+launch_cmd = ['python -m torch.distributed.launch --nproc_per_node 8 ' \
```
Comment (aminsaied) [inline website/docs/cheatsheet/distributed-training.md:273]: ditto - prefer string.split()

File: website/docs/cheatsheet/distributed-training.md [hunk 2]
```diff
@@ -6,524 +6,335 @@ keywords:
...
+               'text-classification/run_glue.py ' \
+               '--model_name_or_path bert-large-uncased-whole-word-masking ' \
+               '--task_name mnli --do_train --do_eval ' \
+               '--max_seq_length 128 --per_device_train_batch_size 8 ' \
+               '--learning_rate 2e-5 --num_train_epochs 3.0 ' \
+               '--output_dir /tmp/mnli_output']
+
+run_config = ScriptRunConfig(source_directory='./src',
```
Comment (aminsaied) [inline website/docs/cheatsheet/distributed-training.md:281]: ditto - args on new line + trailing comma

File: website/docs/cheatsheet/distributed-training.md [hunk 3]
```diff
@@ -6,524 +6,335 @@ keywords:
...
+tf_env = Environment.get(workspace=ws, name=curated_env_name)
 distr_config = TensorflowConfiguration(worker_count=2, parameter_server_count=0)
 
-src = ScriptRunConfig(source_directory=source_dir,
-                      script='train.py',
-                      arguments= ... ,
-                      compute_target= ... ,
-                      environment= ... ,
-                      distributed_job_config=distr_config)
+run_config = ScriptRunConfig(source_directory='./src',
```
Comment (aminsaied) [inline website/docs/cheatsheet/distributed-training.md:304]: ditto: new line + trailing comma

**Review by aminsaied — CHANGES_REQUESTED** [review]

File: website/docs/cheatsheet/distributed-training.md [hunk 4]
```diff
@@ -6,524 +6,335 @@ keywords:
...
-experiment.submit(run_cf)
-```
+To run distributed training with the [DeepSpeed](https://www.deepspeed.ai/) library on Azure ML, do not use DeepSpeed's custom launcher. Instead, configure an MPI job to launch the training job [with MPI](https://www.deepspeed.ai/getting-started/#mpi-and-azureml-compatibility).
 
-:::caution
-To use AzureML mpirun, the base docker image used by the run need to have Open MPI. They are included in [all AzureML default GPU base images](https://github.com/Azure/AzureML-Containers). User is responsible to 
-install one of these when using custom base image. AzureML also provides [curated environment](https://docs.microsoft.com/en-us/azure/machine-learning/resource-curated-environments) for popular frameworks. 
-:::
+Take care of the following:
+* The Azure ML environment contains DeepSpeed and its dependencies, Open MPI, and mpi4py.
```
[the remark this answers is not in the document]
Comment (mx-iao) [reply website/docs/cheatsheet/distributed-training.md:74]: @aminsaied I meant the user should make sure their environment has those dependencies. I'll make the wording more clear. We have that curated environment but I don't know if we should advertise it here since I'm not sure if this was a one-off curated env - let me know what you think

File: website/docs/cheatsheet/distributed-training.md [hunk 5]
```diff
@@ -6,524 +6,335 @@ keywords:
...
+### Launch options
 
-set_environment_variables_for_nccl_backend()
-torch.distributed.init_process_group(backend="nccl")
-local_rank = get_local_rank()
+The Azure ML PyTorch job supports two types of options for launching distributed training:
 
-import torch.distributed as dist
+1. __Per-process-launcher__: The system will launch all distributed processes for the user, with all the relevant information (e.g. environment variables) to set up the process group.
+2. __Per-node-launcher__: The user provides Azure ML with the utility launcher that will get run on each node. The utility launcher will handle launching each of the processes on a given node. Locally within each node, RANK and LOCAL_RANK is set up by the launcher. The **torch.distributed.launch** utility and PyTorch Lightning both belong in this category.
```
[the remark this answers is not in the document]
Comment (mx-iao) [reply website/docs/cheatsheet/distributed-training.md:129]: I linked to the distributed CIFAR example I added yesterday - it shows both options. I'll link to it from both sections

File: website/docs/cheatsheet/distributed-training.md [hunk 6]
```diff
@@ -6,524 +6,335 @@ keywords:
...
+To configure a PyTorch job with a per-node-launcher, do the following:
+1. Provide the `torch.distributed.launch` command to the `command` parameter of the `ScriptRunConfig` constructor. Azure ML will run this command on each node of your training cluster. `--nproc_per_node` should be less than or equal to the number of GPUs available on each node. MASTER_ADDR, MASTER_PORT, and NODE_RANK are all set by Azure ML, so you can just reference the environment variables in the command. Azure ML sets MASTER_PORT to `6105`, but you can pass a different value to the `--master_port` argument of torch.distributed.launch command if you wish. (The launch utility will reset the environment variables.)
+    ```shell
```
[the remark this answers is not in the document]
Comment (mx-iao) [reply website/docs/cheatsheet/distributed-training.md]: updated, let me know if that's not what you were thinking

**Review by aminsaied — APPROVED** [review]
